### PR TITLE
cross join elimination

### PR DESCRIPTION
**Pull request by kokosing**
cross join elimination

### CODE REVIEW

**Comment by kokosing** [comment]
Fixes #4

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
+---
+title: Unnecessary cross join elimination
+tags: optimizer join
+category: internals
+---
+
+Cross join have a bad fame. It is not that nobody likes it all the time. For example
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: have a bad fame -> has a bad reputation

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
+---
+title: Unnecessary cross join elimination
+tags: optimizer join
+category: internals
+---
+
+Cross join have a bad fame. It is not that nobody likes it all the time. For example 
+It is OK to use it from time to time. There are even some queries where there is no other
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: It -> it (w sensie że nie trzeba wielkiej litery)

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+way. All of it is totally acceptable, and nobody would complain if it would be only like
+that. However, cross join has a habit to occur at the least appropriate moment.
+And once it comes, nothing remains the same. Query usually becomes an order of
+magnitude slower, and this not something any of you would like to dream of.
+
+Since the 0.162 and 0.157t (Teradata) version of Presto, there is a feature
+called unnecessary cross join elimination. Simply, it does what it says, if it
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: * dałabym dwukropek po says

Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: (zamiast przecinka)

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+is possible to reorder joins in a way to eliminate cross joins without
+sacrificing the correctness it will happen. This is post is going to uncover the details of it.
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: * przecinek po corectness

Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: This is post -> This post is

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+presto:tiny> show session;
+
+ Name          | Value | Default |  Type   | Description  
+---------------+-------+---------+---------+-------------------------------------------------
+ ...
+ reorder_joins | false | false   | boolean | Experimental: Reorder joins to optimize plan
+ ...
+~~~
+
+There are two ways to turn in on:
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: in -> it

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+~~~
+
+### Why is it not turned on by default?
+
+Unnecessary cross elimination sounds like a great thing to be enabled by
+default, so you may want to ask why it has to be enabled to take the advantage
+of it. I think there are three reasons behind that:
+
+ * As you noticed above, `SHOW SESSION` says it is an experimental feature. 
+It is quite recent feature so there is a slight possibility that it may generate wrong
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: quite recent feature -> a quite recent feature

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+plan for query that may cause wrong results. However, it is tested quite
+vastly and no such thing should happen (although you never know). Anyway,
+anything could a BUG, so should everything be behind the feature toggle?
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: could -> could be

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+Exactly, this reason is not enough to explain why does it need this here.
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: why does it need -> why it needs

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+ * This property may enable other join ordering related optimization which you
+may not want to have. As far I recall there are no such optimization added yet,
+but it the near future it is expected to change.
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: it -> in

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+ * Generally join reordering is tricky. There are queries for which turning on
+this feature may case performance degradation.
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: case -> cause

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+### When cross join happens uninvited? (the use case)
+
+Let take a look at below query and see how it is planned without this great feature:
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: Let -> Let's
below query -> the query below

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+        - TableScan[tpch:tpch:part:sf0.01, originalConstraint = true] ...
+          - ...
+            - TableScan[tpch:tpch:orders:sf0.01, originalConstraint = true] ...
+      - ...
+        - ScanProject[table = tpch:tpch:lineitem:sf0.01, originalConstraint = true] ...
+~~~
+
+As you can see, there are three tables joins together with two conditions. By
+default planner joins them in the order given in the query, so `part` is joined
+with `orders` and then that is joined with `lineitem`, at start all of these
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: at start -> at the beginning

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+joins are cross joins. Then, at the top of them there is `Filter` added with
+the condition from `WHERE` clause (you would not see this from the `EXPLAIN`
+output). After that optimizer notices that there is a `Filter` on top of
+`CrossJoin`. It tries to push it down, thanks to which the first join becomes the `InnerJoin` with two equality conditions. 
+This is all what happened above.
+
+The question is why the optimizer could not push down the join condition any
+further so the other `CrossJoin` remained? Please notice that there are two
+predicates, one joins `lineitem` with `part` (`p.partkey = l.partkey`) and the other joins `lineitem`
+with `orders` (`l.orderkey = o.orderkey`). As you can see are no condition  between `part` and `orders` and
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: condition -> conditions

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+so this join remain as `CrossJoin`.
+
+### How it works?
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: How it works -> How does it work?

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+Exactly, this reason is not enough to explain why does it need this here.
+
+ * This property may enable other join ordering related optimization which you
+may not want to have. As far I recall there are no such optimization added yet,
+but it the near future it is expected to change.
+
+ * Generally join reordering is tricky. There are queries for which turning on
+this feature may case performance degradation.
+
+### When cross join happens uninvited? (the use case)
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: When cross join happens uninvited? -> When does cross join happen uninvited?

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+
+The question is why the optimizer could not push down the join condition any
+further so the other `CrossJoin` remained? Please notice that there are two
+predicates, one joins `lineitem` with `part` (`p.partkey = l.partkey`) and the other joins `lineitem`
+with `orders` (`l.orderkey = o.orderkey`). As you can see are no condition  between `part` and `orders` and
+so this join remain as `CrossJoin`.
+
+### How it works?
+
+Lets enable `reorder_joins` and see what happens.
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: Lets -> Let's

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+   - ...
+     - ScanProject[table = tpch:tpch:orders:sf0.01, originalConstraint = true] ...
+~~~
+
+The most important fact is that join order changed. Now `part` is joined with
+`lineitem` and then this is joined with `orders`. Please notice it matches the
+predicates available in `WHERE` clause, so this time optimizer was able to push
+them dowm and make all the joins to be `InnerJoin`.
+
+Nice, isn't it? But how it was achieved. It is handled by the one of optimizer
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: achieved. -> achieved?

Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: the one -> one

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+rules which creates a join graph. Tables represents vertexes and join
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: vertexes -> może zostać, ale łacińskie pochodzenie sugerowałoby vertices :) tak jak index -> indices, chociaż też można spotkać indexes

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+tables, it chooses the table which appeared on SQL as first and does the join
+of them. In this example only `lineitem` is connected so the decision is simple.
+This is repeated until no more tables can be joined. If some tables
+remain, then it is known that there is no connection to them (no join
+condition) thus in such case there is no way to eliminate all cross joins 
+with this algorithm.
+
+### Summary
+
+This is a nice feature, which in my opinion you should turn it on if you have
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: turn it on -> turn on

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+           |
+l.orderkey = o.orderkey
+           |
+           |
+           +
+        orders o
+~~~
+
+Now this rule takes the table which appeared on SQL query as first (in this
+example `part`) and see what tables are connected to it. Among the connected
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: see -> sees

File: _posts/2017-01-28-unnecessary-cross-join-elimination.md [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+remain, then it is known that there is no connection to them (no join
+condition) thus in such case there is no way to eliminate all cross joins 
+with this algorithm.
+
+### Summary
+
+This is a nice feature, which in my opinion you should turn it on if you have
+not done it already. I have heard that some databases display a warning (or
+even raise an error if you configure such option) when a cross join is used,
+that way user have a better notion when cross join was used.
```
Comment (nat925) [inline _posts/2017-01-28-unnecessary-cross-join-elimination.md]: user -> users

**Pull request merged**